lib/vrl/compiler/src/type_def.rs

```diff
@@ -1172,6 +1172,141 @@ impl TypeDef {
             kind: newkind,
         }
     }
+
+    fn remove_segment<'a, I>(kind: &KindInfo, mut path: std::iter::Peekable<I>) -> KindInfo
+    where
+        I: std::iter::Iterator<Item = &'a SegmentBuf> + Clone,
+    {
+        match kind {
+            KindInfo::Unknown => {
+                // The kind is unknown and thus there is nothing to remove.
+                kind.clone()
+            }
+            KindInfo::Known(kinds) => KindInfo::Known(
+                kinds
+                    .iter()
+                    .map(|kind| match (kind, path.next(), path.peek()) {
+                        (TypeKind::Object(object), Some(SegmentBuf::Field(_)), Some(_)) => {
+                            TypeKind::Object(
+                                object
+                                    .iter()
+                                    .map(|(field, kindinfo)| {
+                                        (
+                                            field.clone(),
+                                            Self::remove_segment(kindinfo, path.clone()),
+                                        )
+                                    })
+                                    .collect(),
+                            )
+                        }
+
+                        (TypeKind::Object(object), Some(SegmentBuf::Field(fieldname)), None) => {
+                            TypeKind::Object(
+                                object
+                                    .iter()
+                                    .filter_map(|(field, kindinfo)| match field {
+                                        Field::Field(name) if name == fieldname.as_str() => None,
+                                        _ => Some((field.clone(), kindinfo.clone())),
+                                    })
+                                    .collect(),
+                            )
+                        }
```

Comment:
Whats the difference between these two match arms?

Reply:
I guess my questions is generally between the `thing(_)` vs `thing(specific)`

Reply:
I think in the first one, the important part is matching the `Some` on the 3rd element; the field didn't need binding.

Whereas the second arm is `None` in the 3rd element, and the fieldname is needed.

Reply:
Oh duh - I misread that and thought it was just matching on the end bit. Not enough ☕

Reply:
It's actually wrong!

The first arm means that we aren't at the end of the path and so we need to recurse down, the second arm we are at the end of the path and need to remove the field.

The problem is it is recursing down all the the fields of the object, whereas it should only recurse down the one indicated by the current field.

Well spotted, you could see something wasn't quite right!